docs: add tutorial to distribute artifact to zot registry (#187)
oras-project/oras-www: docs/Tutorials/zot-registry-tutorial.mdx

latest: Pulling from project-zot/zot-linux-amd64
a7ca0d9ba68f: Pull complete 
fe5ca62666f0: Pull complete 

<!-- truncated -->

after":"2h0m0s","goroutine":24,"caller":"zotregistry.io/zot/pkg/extensions/extension_search.go:160","time":"2023-05-20T10:50:45.892076536Z"}
{"level":"debug","component":"scheduler","worker":3,"goroutine":24,"caller":"zotregistry.io/zot/pkg/scheduler/scheduler.go:105","time":"2023-05-20T10:50:45.892089946Z","message":"scheduler: finished task"}
```

### Phase 2: Distribution of OCI Artifacts to Zot Registry Using Oras CLI

Please open a new terminal window for the next phase.

**Step 1: Create a sample file** 

```
echo "hello world" > artifact.txt
```

**Step 2: Pushing artifacts to the zot repository**

```
oras push --plain-http localhost:5000/hello-artifact:v1 \
        --artifact-type application/vnd.acme.rocket.config \
        artifact.txt:text/plain -d -v
```

```
Uploading a948904f2f0f artifact.txt
Uploaded  a948904f2f0f artifact.txt
Pushed [registry] localhost:5000/hello-artifact:v1
Digest: sha256:bcdd6799fed0fca0eaedfc1c642f3d1dd7b8e78b43986a89935d6fe217a09cee
```

After pushing the artifact, it can be seen in the zot user interface:

> ![](https://hackmd.io/_uploads/SJdQHUKSn.png)

Reviewer: Submit the picture to the version library, and the external link will become invalid in the future

Reviewer 2: Missed this, the pace for images is `static/img`

Author: Done, thank you!

---

> title: Tutorial distribute an OCI artifact to Zot Registry using ORAS CLI

Reviewer 2: I wonder if this should just be called something along the lines of introduction to OCI registries and artifacts.

Author: I think we'll add more use cases so we might have to mention zot. If not, I agree, we can remove zot from the title.

Reviewer 3: I'd also recommend that we make any guidance to be as solution/vendor agnostic as possible

Reviewer 4: Agreed, the title can say "... to OCI registries" and in the following text, make `zot` an example so readers have something to follow along and try.

Reviewer 4: Also the term "artifact" (over time) is loaded/controversial/confusing. I think you want to refer to: https://github.com/opencontainers/image-spec/pull/1068 and related corpus.

Author: Thank you, everyone! I have updated the title now and it does not have zot in it.

Author: @rchincha if I understood correctly, I should be using the term "OCI image" here, right?

Reviewer 4: "images and artifacts"?

Author: Done 👍

sidebar_position: 0
---

> # Tutorial to Install Zot Registry and Distribute OCI Artifacts to Zot Registry using ORAS CLI

Reviewer 5: ```suggestion
# Manage OCI Artifacts in Zot Registry with ORAS
```

Before we tackle how to distribute an OCI artifact to an OCI registry, let us first take a look at what we mean by distributing OCI artifacts, why we need it and what OCI registries are. 

> OCI registries are like a single location where container images are stored and made available to everyone. Distributing OCI artifacts refers to when artifacts are pushed to these registries. These artifacts can then be pulled (or retrieved) by someone else and used in their environments. These images often help with packaging and distributing software applications.

Reviewer 2: container images and other artifacts

> We will be using zot registry in this guide. Zot registry is an OCI-native container registry for distributing container images and OCI artifacts.

Reviewer 2: Is that what they say OCI-native or do they say OCI compliant?

Author: It's OCI-native on their website that's why I've used that term.

Reviewer 4: because ... 

zot: a production-ready vendor-neutral OCI image registry - images stored in [OCI image format](https://github.com/opencontainers/image-spec), [distribution specification](https://github.com/opencontainers/distribution-spec) on-the-wire, that's it!

That said, it also supports oras artifacts/referrers until OCI artifacts land (undecided at this time though)

Author: Thank you @rchincha!

> We will be using zot registry in this guide. Zot registry is an OCI-native container registry for distributing container images and OCI artifacts.

Reviewer 6: I would also call out that this setup is going to be plain http and only for testing purposes as this is insecure. For TLS setup follow - https://github.com/project-zot/zot/blob/main/examples/README.md#network

Author: Thank you, I have added it to a note at the end of the tutorial. Is adding the link enough or should I cover more in the tutorial?

Get the application URL by running these commands:
  export NODE_PORT=$(kubectl get --namespace default -o jsonpath="{.spec.ports[0].nodePort}" services zot)
  export NODE_IP=$(kubectl get nodes --namespace default -o jsonpath="{.items[0].status.addresses[0].address}")
  echo http://$NODE_IP:$NODE_PORT
```

Step 5: To verify the installation

```
export NODE_PORT=$(kubectl get --namespace default -o jsonpath="{.spec.ports[0].nodePort}" services zot)
export NODE_IP=$(kubectl get nodes --namespace default -o jsonpath="{.items[0].status.addresses[0].address}")
echo http://$NODE_IP:$NODE_PORT
```

Your output may be different but it should look like this:
```
http://192.168.49.2:30276
```

If you redirect to the above link, the screen will show the following:

![](https://hackmd.io/_uploads/SJT67IYH3.png)

### Phase 2: Distribution of OCI Artifacts to Zot Registry Using Oras CLI

Please open a new terminal window for the next phase.

**Note:** If you have used method 1 for the installation, you will need to replace the IP address and port with `localhost:5000`.

**Step 1: Create a sample file** 

```
echo "hello world" > artifact.txt
```

**Step 2: Authenticate with zot registry**

> It is to be noted that in case of zot, authentication is disabled by default. Still, if you wish to authenticate, you can use [`oras login`](https://oras.land/docs/CLI_Reference/oras_login) to do so.

Author: I've made changes in this commit. Thank you!

> It is to be noted that in case of zot, authentication is disabled by default. Still, if you wish to authenticate, you can either use `docker login` or [`oras login`](https://oras.land/docs/CLI_Reference/oras_login) to do so.

Reviewer 2: ```suggestion
It is to be noted that in case of zot, authentication is disabled by default. Still, if you wish to authenticate, you can use [`oras login`](https://oras.land/docs/CLI_Reference/oras_login) to do so.
```
This is an oras tutorial, might as well cover the oras way

Downloaded  bcdd6799fed0 application/vnd.oci.image.manifest.v1+json
Pulled [registry] 192.168.49.2:30276/hello-artifact:v1
Digest: sha256:bcdd6799fed0fca0eaedfc1c642f3d1dd7b8e78b43986a89935d6fe217a09cee
```

**Step 5: To attach files to an existing artifact**

First, let's create another sample file to add,

```
echo "hello world" > hi.txt
```

You can use the command below to attach `hi.txt` to the artifact we pushed above:

```
oras attach --plain-http --artifact-type doc/example 192.168.49.2:30276/hello-artifact:v1 hi.txt
```

```
Exists    a948904f2f0f hi.txt
Attached to [registry] 192.168.49.2:30276/hello-artifact@sha256:327db68f73d0ed53d528d927a6703c00739d7c1076e50762c3f6641b51b76fdc
Digest: sha256:bcdd6799fed0fca0eaedfc1c642f3d1dd7b8e78b43986a89935d6fe217a09cee
```

**Step 6: To check the referrers of the manifest in our zot registry**

```
oras discover --plain-http 192.168.49.2:30276/hello-artifact:v1
```

```
Discovered 1 artifact referencing v1
Digest: sha256:327db68f73d0ed53d528d927a6703c00739d7c1076e50762c3f6641b51b76fdc

Artifact Type   Digest
doc/example     sha256:bcdd6799fed0fca0eaedfc1c642f3d1dd7b8e78b43986a89935d6fe217a09cee
```

> **Note:** We have used the flag `--plain-http` in the steps because we want to allow insecure connections to registry without SSL check.

Author: Thank you, I have changed that in the note.

### Conclusion

You can now successfully push OCI artifacts to your zot registry! 

> Distributing OCI artifacts to OCI registries aids in collaboration and code sharing. It enables teams to acquire and use standardized container images. This is why it is considered to play a crucial role in maintaining consistency among teams.

Reviewer 2: ```suggestion
Distributing OCI artifacts to OCI registries aids in collaboration and code sharing. It enables teams to acquire and use through a standardized artifact interface. This is why it is considered to play a crucial role in maintaining consistency among teams.
```
I'm not entirely happy with that suggestion, but maybe something along those lines

Author: Okay, I'll change it to that for now and try to come up with something along those lines

> Distributing OCI artifacts to OCI registries aids in collaboration and code sharing. It enables teams to acquire and use through a standardized artifact interface. This is why it is considered to play a crucial role in maintaining consistency among teams.

Author: Yes, I'll work on the conclusion and update it soon.

> Distributing OCI artifacts to OCI registries aids in collaboration and code sharing. It enables teams to acquire and use through a standardized artifact interface. This is why it is considered to play a crucial role in maintaining consistency among teams.

Author: @sabre1041, the conclusion has been changed a little. Does it look better now or would you like me to add more content to it?
